lmdeploy/pytorch/backends/default/moe.py

```diff
@@ -7,21 +7,35 @@
 class DefaultSoftmaxTopKImpl(SoftmaxTopKImpl):
     """RMS norm implementation api."""
 
-    def __init__(self, top_k: int, dim: int = -1):
+    def __init__(self, top_k: int, dim: int = -1, n_groups: int = -1):
         self.top_k = top_k
         self.dim = dim
+        self.n_groups = n_groups
 
     def forward(self, x: torch.Tensor):
         """forward."""
         routing_weights = torch.softmax(x, dim=self.dim, dtype=torch.float32)
-        topk_weights, topk_ids = torch.topk(routing_weights, self.top_k, dim=self.dim)
+        if self.n_groups > 0:
+            assert self.top_k % self.n_groups == 0, f'{self.top_k} cannot be divided by {self.n_groups}'
+            assert routing_weights.shape[
+                -1] % self.n_groups == 0, f'{routing_weights.shape[-1]} cannot be divided by {self.n_groups}'
+            per_group_top_k = self.top_k // self.n_groups
+            group_size = routing_weights.shape[-1] // self.n_groups
+            group_offsets = (torch.arange(self.n_groups, device=routing_weights.device) * group_size).view(
```

Comment:
Since  n_groups is fixed, we can create a function with lru_cache

```python
@functools.lru_cache
def get_group_offsets(n_groups, group_size, device):
     return ....
```

So we can reuse the same group_offsets